### PR TITLE
Add generic pagination for the grapqhl request

### LINKED ISSUE #627: Make GraphQL pagination generic
### Description
Make graphQL pagination generic to have a re-usable solution; check https://github.com/commercetools/commercetools-sync-java/pull/619#discussion_r521917982

### PR DESCRIPTION
**Pull request by ahmetoz**
Fixes #627 

Note: `FetchCustomObjectsGraphQlRequest` is the request I will need on another task, so this task is opened to simplify the PR of other work (currently Is #651).

### CODE REVIEW

File: src/main/java/com/commercetools/sync/commons/models/GraphQlBaseRequest.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package com.commercetools.sync.commons.models;
+
+import io.sphere.sdk.client.SphereRequest;
+
+public interface GraphQlBaseRequest<T> extends SphereRequest<T>
```
Comment (lojzatran) [inline src/main/java/com/commercetools/sync/commons/models/GraphQlBaseRequest.java:5]: Can you describe what T should be?

Comment (lojzatran) [reply src/main/java/com/commercetools/sync/commons/models/GraphQlBaseRequest.java:5]:
```suggestion
public interface GraphQlBaseRequest<T extends GraphQlBaseResult<?>> extends SphereRequest<T>
```

**Review by lojzatran — COMMENTED** [review]
LGTM 🎉  I just have some small suggestions.

**Review by lojzatran — CHANGES_REQUESTED** [review]
test

File: src/main/java/com/commercetools/sync/commons/helpers/CommonGraphQlRequestImpl.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import io.sphere.sdk.http.HttpMethod;
+import io.sphere.sdk.http.HttpResponse;
+import io.sphere.sdk.json.SphereJsonUtils;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import static java.lang.String.format;
+
+public abstract class CommonGraphQlRequestImpl<T> implements GraphQlBaseRequest<T> {
```
[the remark this answers is not in the document]
Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/commons/helpers/CommonGraphQlRequestImpl.java]: https://github.com/commercetools/commercetools-sync-java/pull/649/commits/627eb8225b4a078062ec1e66170a086a004aed6c

File: src/main/java/com/commercetools/sync/commons/helpers/CommonGraphQlRequestImpl.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        JsonNode result = rootJsonNode.get("data").get(resourceName);
+        return SphereJsonUtils.readObject(result, clazz);
+    }
+
+    /**
+     * This method builds a string matching the required format needed in the CTP graphql API.
+     *
+     * @return a string representing a graphql query
+     */
+    protected abstract Object buildQueryString();
```
[the remark this answers is not in the document]
Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/commons/helpers/CommonGraphQlRequestImpl.java]: https://github.com/commercetools/commercetools-sync-java/pull/649/commits/a32c0fd4dc5f8b26b282cbf75dac776002183d3a

File: src/main/java/com/commercetools/sync/commons/models/FetchCustomObjectsGraphQlRequest.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+public class FetchCustomObjectsGraphQlRequest extends CommonGraphQlRequestImpl<ResourceKeyIdGraphQlResult> {
+
+    private final String container;
+    private final Instant lastModifiedAt;
+    private final String resourceName = GraphQlQueryResources.CUSTOM_OBJECTS.getName();
+
+    /**
+     * A SphereRequest implementation to allow {@link SphereClient} to execute graphQL queries on CTP. It provides a
+     * POST request to the CTP graphql API containing body to fetch a set of keys matching given container of a resource
+     * defined in endpoint parameter.
```
[the remark this answers is not in the document]
Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/commons/models/FetchCustomObjectsGraphQlRequest.java]: https://github.com/commercetools/commercetools-sync-java/pull/649/commits/2ad04e5d7bd27a2eeca68b9fcb91f054bd3d9f0a

File: src/main/java/com/commercetools/sync/commons/models/GraphQlBaseRequest.java (hunk 1 above)
Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/commons/models/GraphQlBaseRequest.java]: https://github.com/commercetools/commercetools-sync-java/pull/649/commits/d87506aff7ee6ab970286c0b42ba26ca2ab9c0f9

**Comment by ahmetoz** [comment]
@lojzatran really good review, thanks!

**Comment by salander85** [comment]
@ahmetoz Nice work, looks good +1

**Review by lojzatran — APPROVED** [review]

**Pull request merged**